File: ColoringBookApp/Features/Canvas/UndoStack.swift
import Foundation
import UIKit
import PencilKit

enum CanvasAction {
    case fill(label: UInt32, indices: [Int], previous: [UInt32], newColor: UIColor)
    case brush(previous: PKDrawing, next: PKDrawing)
    case sticker(StickerOp)
    case clear(previous: PaintLayerSnapshot)
}

struct PaintLayerSnapshot {
    let pixels: [UInt32]
}

enum StickerOp {
    case add(StickerInstance)
    case remove(StickerInstance)
    case transform(id: UUID, before: StickerTransform, after: StickerTransform)
}

final class UndoStack {
    private var undoActions: [CanvasAction] = []
    private var redoActions: [CanvasAction] = []

    var canUndo: Bool { !undoActions.isEmpty }
    var canRedo: Bool { !redoActions.isEmpty }

    func push(_ action: CanvasAction) {
        undoActions.append(action)
        redoActions.removeAll()
    }

    func popUndo() -> CanvasAction? {
        guard let a = undoActions.popLast() else { return nil }
        redoActions.append(a)
        return a
    }

    func popRedo() -> CanvasAction? {
        guard let a = redoActions.popLast() else { return nil }
        undoActions.append(a)
        return a
    }

    func clear() {
        undoActions.removeAll()
        redoActions.removeAll()
    }
}

File: web/js/canvas/undo-stack.js
// Mixed undo stack covering bucket fills, brush strokes, and stickers.
// Each action carries the data needed to undo and redo without re-rendering
// the whole stage.

export class UndoStack {
  constructor() {
    this.undoActions = [];
    this.redoActions = [];
  }
  get canUndo() { return this.undoActions.length > 0; }
  get canRedo() { return this.redoActions.length > 0; }

  push(action) {
    this.undoActions.push(action);
    this.redoActions.length = 0;
  }
  popUndo() {
    const a = this.undoActions.pop();
    if (a) this.redoActions.push(a);
    return a;
  }
  popRedo() {
    const a = this.redoActions.pop();
    if (a) this.undoActions.push(a);
    return a;
  }
  clear() {
    this.undoActions.length = 0;
    this.redoActions.length = 0;
  }
}

// Action shapes (loosely typed):
// { type: "fill",   indices: Uint32Array, prev: Uint32Array, next: Uint32 }
// { type: "brush",  prev: ImageData,      next: ImageData }
// { type: "sticker-add",    sticker: {...} }
// { type: "sticker-remove", sticker: {...} }
// { type: "sticker-move",   id, before: {x,y,scale,rot}, after: {...} }
// { type: "clear",  prev: Uint32Array, prevBrush: ImageData, prevStickers: [...] }
